File: python/baekjoon/lv.gold/3/트리의지름.py
# link : https://www.acmicpc.net/problem/1167
# Lv : gold 3
# Category : DFS, 트리

# 이 문제는 트리가 주어지고 각 트리를 연결하는 간선의 가중치가 있는 모습이다.
# 시간초과에 대한 문제를 해결하기 위해서 사용한 방법은 최대값이 나올만한 노드를 미리 찾아서 해당 노드만 탐색을 진행하는 방법이다. 
# 트리의 노드들은 모두 연결이 되어있으며 한 노드에서 다른 노드까지의 가중치의 합을 구해서 비교해보면 최대값이 나올 수 있는 노드를 구할 수 있다.
# 👉 여기서 한 노드만을 통해서 구한 최대값을 그대로 사용하면 되지 않는가? 라는 의문이 떠오를 수 있지만, 만약 해당 노드의 부모노드가 존재하게 된다면 이 값은 최종적인 답이 될 수 없다.
# 따라서 findMaxIndex를 통해서 dfs 탐색을 할 노드를 찾고 탐색을 진행해주면 된다.

import sys

answer = 0
N = int(sys.stdin.readline())
tree = {i : [] for i in range(1, N+1)}
check = [-1]*(N+1)

for i in range(N):
  li = list(map(int, sys.stdin.readline().split()))
  root = li[0]
  li.pop()
  while len(li) > 1:
    dist = li.pop()
    node = li.pop() 
    tree[root].append((node, dist))


def findMaxIndex(node, check):
  for destination, dist in tree[node]:
    if check[destination] == -1:
      check[destination] = check[node] + dist  
      findMaxIndex(destination, check)

def dfs(node, r, visited):
  global answer
  for destination, dist in tree[node]:
    if destination not in visited:
      visited.add(node)
      dfs(destination, r+dist, visited)
      visited.remove(node)
    else:
      answer = max(answer, r)
    

findMaxIndex(1, check)
dist = 0
startNode = 1 
for i in range(1, N+1):
  if dist < check[i]:
    dist = check[i]
    startNode = i


dfs(startNode, 0, set())

print(answer)

File: java/baekjoon/lv.gold/4/트리의지름.java
/* 
    link : https://www.acmicpc.net/problem/1967
    Lv : gold 4
    Category : BFS
    
    루트 노드가 정해져 있는 트리에서 각 가중치를 이용한 트리의 최대 지름을 구하는 문제이다.
    우선 BFS를 통해서 루트 노드로부터 있는 가장 먼 노드를 찾는다.
    그 노드로 부터 다른 단일노드까지의 거리를 BFS를 통해서 탐색하여 최대 가중치의 합을 구했다.
*/

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Queue;

class Point {
	int node;
	int distance;

	public Point(int node, int distance) {
		this.node = node;
		this.distance = distance;
	}
}

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
		int nodes = Integer.parseInt(bf.readLine());
		HashMap<Integer, ArrayList<int[]>> graph = new HashMap<>();
		for (int i = 0; i < nodes; i++) {
			graph.put(i + 1, new ArrayList());
		}
		for (int i = 0; i < nodes - 1; i++) {
			String[] edgeCases = bf.readLine().split(" ");
			int parentNode = Integer.parseInt(edgeCases[0]);
			int childNode = Integer.parseInt(edgeCases[1]);
			int weight = Integer.parseInt(edgeCases[2]);
			graph.get(parentNode).add(new int[] {childNode, weight});
			graph.get(childNode).add(new int[] {parentNode, weight});
		}
		if (nodes == 1) {
			System.out.println(0);
			return;
		}
		Point maxPointFromRoot = findMaxPoint(graph, 1, nodes);
		Point maxPointFromLeaf = findMaxPoint(graph, maxPointFromRoot.node, nodes);
		System.out.println(maxPointFromLeaf.distance);
	}

	private static Point findMaxPoint(HashMap<Integer, ArrayList<int[]>> graph, int startNode, int nodes) {
		boolean[] visited = new boolean[nodes + 1];
		for (int i = 0; i < nodes; i++) {
			visited[i + 1] = false;
		}
		Queue<Point> queue = new LinkedList<>();
		queue.add(new Point(startNode, 0));
		Point maxPoint = new Point(0, 0);

		while (!queue.isEmpty()) {
			Point point = queue.poll();
			if (graph.get(point.node).size() == 1 && point.node != startNode) {
				if (maxPoint.distance < point.distance) {
					maxPoint = point;
				}
				continue;
			}
			for (int[] array : graph.get(point.node)) {
				int nextNode = array[0];
				int weight = array[1];
				if (visited[nextNode] == true) {
					continue;
				}
				queue.add(new Point(nextNode, point.distance + weight));
			}
			visited[point.node] = true;
		}
		return maxPoint;
	}
}
